consensus/crdt/consensus.go

```diff
@@ -0,0 +1,410 @@
+	logging "github.com/ipfs/go-log"
+	rpc "github.com/libp2p/go-libp2p-gorpc"
+	host "github.com/libp2p/go-libp2p-host"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	peer "github.com/libp2p/go-libp2p-peer"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+)
+
+var logger = logging.Logger("crdt")
+
+var (
+	blocksNs = "b" // blockstore namespace
+)
+
+// Common variables for the module.
+var (
+	ErrNoLeader = errors.New("crdt consensus component does not provide a leader")
+	ErrRmPeer   = errors.New("crdt consensus component cannot remove peers")
+)
+
+// Consensus implement ipfscluster.Consensus and provides the facility to add
+// and remove pins from the Cluster shared state. It uses a CRDT-backed
+// implementation of go-datastore (go-ds-crdt).
+type Consensus struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	config *Config
+
+	host host.Host
+
+	store     ds.Datastore
+	namespace ds.Key
+
+	state state.State
+	crdt  *crdt.Datastore
+
+	dht    *dht.IpfsDHT
+	pubsub *pubsub.PubSub
+
+	rpcClient *rpc.Client
+	rpcReady  chan struct{}
+	readyCh   chan struct{}
+
+	shutdownLock sync.RWMutex
+	shutdown     bool
+}
+
+// New creates a new crdt Consensus component. The given PubSub will be used to
+// broadcast new heads. The given thread-safe datastore will be used to persist
+// data and all will be prefixed with cfg.DatastoreNamespace.
+func New(
+	host host.Host,
+	dht *dht.IpfsDHT,
+	pubsub *pubsub.PubSub,
+	cfg *Config,
+	store ds.Datastore,
+) (*Consensus, error) {
+	err := cfg.Validate()
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	css := &Consensus{
+		ctx:       ctx,
+		cancel:    cancel,
+		config:    cfg,
+		host:      host,
+		dht:       dht,
+		store:     store,
+		namespace: ds.NewKey(cfg.DatastoreNamespace),
+		pubsub:    pubsub,
+		rpcReady:  make(chan struct{}, 1),
+		readyCh:   make(chan struct{}, 1),
+	}
+
+	go css.setup()
+	return css, nil
+}
+
+func (css *Consensus) setup() {
+	select {
+	case <-css.ctx.Done():
+		return
+	case <-css.rpcReady:
+	}
+
+	// Validate pubsub messages for our topic (only accept
+	// from trusted sources)
+	err := css.pubsub.RegisterTopicValidator(
+		css.config.ClusterName,
+		func(ctx context.Context, p peer.ID, msg *pubsub.Message) bool {
+			// This is where peer authentication will go.
+			return true
+		},
+	)
+	if err != nil {
+		logger.Errorf("error registering topic validator: %s", err)
+	}
+
+	var blocksDatastore ds.Batching
+	blocksDatastore = namespace.Wrap(css.store, css.namespace.ChildString(blocksNs))
+
+	ipfs, err := ipfslite.New(
+		css.ctx,
+		blocksDatastore,
+		css.host,
+		css.dht,
+		&ipfslite.Config{
+			Offline: false,
+		},
+	)
+	if err != nil {
+		logger.Errorf("error creating ipfs-lite: %s", err)
+		return
+	}
+
+	dagSyncer := newLiteDAGSyncer(css.ctx, ipfs)
+
+	broadcaster, err := crdt.NewPubSubBroadcaster(
+		css.ctx,
+		css.pubsub,
+		css.config.ClusterName,
+	) // subscription name
```

Comment:
comment should be at the end of the line above

Reply:
ok